src/implements/thread_pool.cc

```diff
@@ -0,0 +1,60 @@
+#include "include/thread_pool.hpp"
+#include <iostream>
+
+ThreadPool::ThreadPool(uint16_t poolSize) {
+  auto max_concurrency = std::thread::hardware_concurrency();
+  uint16_t size_ = poolSize > 0 ? poolSize : 1;
+  if(size_ > max_concurrency * 5)
+    size_ = max_concurrency * 5;
+  available_.store(size_);
+  processing_.store(true);
+  for (int i = 0; i < size_; ++i) {
+    workers_.emplace_back(std::bind(&ThreadPool::initThread, this));
+  }
+}
+
+ThreadPool::~ThreadPool() {
+  if(processing_.load()) {
+    abort();
+  }
+}
+
+void ThreadPool::initThread() {
+  while (processing_.load()) {
+    std::function<void()> task;
+    {
+      std::unique_lock<std::mutex> lock(mutex_);
+      cv_.wait(lock, [this] {
+        // if pool is running and exists pending jobs, stop blocking thread
+        bool condition = !(processing_.load() && jobs_.empty());
+        return condition;
+      });
+      if (!processing_.load() && jobs_.empty()) return;
```

Comment:
If processing_.load() is true and jobs_.empty() is true, here should return?
Because there is no job to be moved to task at line 34. So the if condition should be 
(!processing_.load() || jobs_.empty()).
Correct me if I'm wrong.

Reply:
是的，
Line 29 可修改為 !jobs_.empty() || !processing_.load() 來短路求值，
可先判定若 jobs 不為空，結束 wait (存在jobs)；若 jobs為空但是 processing_ 為 false，同樣結束 wait (避免dead lock)

Line 34 則修改為 (!processing_.load() || jobs_.empty())，
若processing_ 為 false直接退出；processing_為true 且 jobs_為空亦退出 thread
僅在 jobs_不為空且 processing_ 為 true才繼續執行